src/add.cpp

```diff
@@ -330,8 +330,9 @@ RCP<const Basic> sub(const RCP<const Basic> &a, const RCP<const Basic> &b)
     return add(a, mul(minus_one, b));
 }
 
-RCP<const Basic> add_expand(const RCP<const Add> &self)
+RCP<const Basic> add_expand(const RCP<const Basic> &self_)
 {
+    const RCP<const Add> self = rcp_static_cast<const Add>(self_);
```

Comment:
I don't like that the casting happens in these routines. It should happen before `add_expand` is called, but I can't figure out how to do that fast yet.